fendermint/app/src/app.rs

```diff
@@ -666,26 +706,40 @@ where
             "process proposal"
         );
         let txs: Vec<_> = request.txs.into_iter().map(|tx| tx.to_vec()).collect();
+        let size_txs = txs.iter().map(|tx| tx.len()).sum::<usize>();
         let num_txs = txs.len();
 
-        let accept = self
-            .interpreter
-            .process(self.chain_env.clone(), txs)
-            .await
-            .context("failed to process proposal")?;
-
-        emit!(ProposalProcessed {
-            is_accepted: accept,
-            block_height: request.height.value(),
-            block_hash: request.hash.to_string().as_str(),
-            num_txs,
-            proposer: request.proposer_address.to_string().as_str()
+        let process_result = self.interpreter.process(self.chain_env.clone(), txs).await;
+
+        emit(BlockProposalReceived {
+            height: request.height.value(),
+            hash: HexEncodableBlockHash(request.hash.into()),
+            size: size_txs,
+            tx_count: num_txs,
+            validator: request.proposer_address.to_string().as_str(),
         });
 
-        if accept {
-            Ok(response::ProcessProposal::Accept)
-        } else {
-            Ok(response::ProcessProposal::Reject)
+        match process_result {
+            Ok(_) => {
+                emit(BlockProposalAccepted {
+                    height: request.height.value(),
+                    hash: HexEncodableBlockHash(request.hash.into()),
+                    size: size_txs,
+                    tx_count: num_txs,
+                    validator: request.proposer_address.to_string().as_str(),
+                });
+                Ok(response::ProcessProposal::Accept)
+            }
+            Err(e) => {
+                emit(BlockProposalRejected {
+                    height: request.height.value(),
+                    size: size_txs,
+                    tx_count: num_txs,
+                    validator: request.proposer_address.to_string().as_str(),
+                    reason: e.to_string().as_str(),
+                });
```

Comment:
Similar to the above, we could merge these types into a single `BlockProposalEvaluated` type.

Reply:
Yes. I have changed it to be more DRY friendly!